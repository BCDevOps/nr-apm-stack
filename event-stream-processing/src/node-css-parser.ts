import {injectable} from 'inversify';
import {Parser} from './parser.isvc';
import * as lodash from 'lodash';
import {renameField} from './shared/rename-field';

/**
* This class can parse logs that are received in a JOSN format
* to make them more suited to indexing in a log database
* This class has been used to parse logs from a NodeJS/Express web application
* output with an Express Winston logger (https://www.npmjs.com/package/express-winston)
* and forwarded from the application by Fluent-bit
* @author The Common Services Showcase team <NR.CommonServiceShowcase@gov.bc.ca>
* @author Tim Csaky <tim.csaky@gov.bc.ca>
* @summary A parser for converting application logs into an ECS compatible format
*/
@injectable()
export class NodeCssParser implements Parser {
  /**
   * Checks various properties of the incoming log (eg: event.kind and event.dataset)
   * to determine if it should be parsed by this parser
   * @param {any} record
   */
  matches(record: any): boolean {
    return record.event?.kind === 'event' &&
      // only parse 'http' (access) logs for now.
      // If parsing error and info logs, change to:
      // lodash.includes(record?.event.dataset, 'node.css');
      lodash.includes(record?.event.dataset, 'node.css.http');
  }
  /**
   * Applies formatting changes to the incoming log
   * to make it match the ELastic Common Schema
   * see: https://www.elastic.co/guide/en/ecs/current/ecs-reference.html#_what_is_ecs
   * @param {any} record
   */
  apply(record: any): void {
    // parse logs from Fluent-bit
    if (lodash.get(record, 'agent.type', '') === 'fluentbit') {
      record.clientIp = this.chooseIpvFormat(record.clientIp);
      renameField(record, 'clientIp', 'client.ip');
      renameField(record, 'contentLength', 'http.response.body.bytes');
      renameField(record, 'hostname', 'host.hostname');
      renameField(record, 'httpVersion', 'http.version');
      record.hostIp = this.chooseIpvFormat(record.hostIp);
      renameField(record, 'hostIp', 'host.ip');
      renameField(record, 'log', 'event.original');
      renameField(record, 'logFileOffset', 'log.file.offset');
      renameField(record, 'logFilePath', 'log.file.path');
      renameField(record, 'method', 'http.request.method');
      renameField(record, 'namespace', 'host.namespace');
      renameField(record, 'path', 'url.path');
      record.query = this.showQueryParamsAsString(record.query);
      renameField(record, 'query', 'url.query');
      renameField(record, 'responseTime', 'http.response.time');
      renameField(record, 'statusCode', 'http.response.status_code');
      renameField(record, 'timestamp', '@timestamp');
      renameField(record, 'userAgent', 'user_agent.original');
    }
  }

  /**
 * NodeJS app is outputting the record.hostIp as an IPV6 formatted ip address.
 * eg: '::ffff:10.97.6.1'.  * '::ffff:' is a designated as the IPv4 to IPv6 subnet prefix
 * to show IPv4 addresses in an IPV6 format.
 * see: https://whatismyipaddress.com/work-arounds
 * In an effort to unify ECS ip addresses we can revert to a regular IPV4 address
 * (eg: '10.97.6.1') if it contains this mechanism
 * @param {string} value - the original record.ip field
 */
  chooseIpvFormat(value: string) {
    return value.startsWith('::ffff:') ? value.replace(/ /g, '').split(':').slice(-1).pop() : value;
  }

  /**
   * serialize query paramaters
   * from object properties back to a query string
   */
  showQueryParamsAsString(query: any) {
    const queryString = Object.keys(query)
      .map((key) => `${key}=${query[key]}`)
      .join('&');
    return queryString;
  }
}
