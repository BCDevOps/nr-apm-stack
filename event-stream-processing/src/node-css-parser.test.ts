import { NodeCssParser } from './node-css-parser';

test('basic', () => {
  const filter = new NodeCssParser();

  /* eslint-disable */

  // record from fluent-bit looks like this:
  /*

  [
    1
  ] app.http: [
    1626304625.036902482,
    {
      "statusCode"=>200,
      "userAgent"=>"Mozilla/5.0",
      "path"=>"/api/v1/forms",
      "message"=>"GET /api/v1/forms 200 624ms",
      "logFilePath"=>"/var/log/app.log",
      "logFileOffset"=>759,
      "hostname"=>"chefs-app-123",
      "timestamp"=>"2021-07-14T23:17:05.036Z",
      "product"=>"chefs",
      "namespace"=>"123-dev",
      "azp"=>"chefs-frontend",
      "level"=>"http",
      "contentLength"=>"323",
      "agent"=>{
        "type"=>"fluentbit"
      },
      "responseTime"=>624,
      "httpVersion"=>"1.1",
      "event"=>{
        "type"=>"http",
        "module"=>"node.css",
        "kind"=>"event",
        "dataset"=>"node.css.http"
      },
      "ip"=>"::ffff:10.97.8.1",
      "log"=>"{"responseTime":624,"azp":"chefs-frontend","contentLength":"323","httpVersion":"1.1","ip":": :ffff: 10.97.8.1","method":"GET","path":"/api/v1/forms","statusCode":200,"userAgent":"Mozilla/5.0","level":"http","message":"GET /api/v1/forms 200 624ms","timestamp":"2021-07-14T23: 17: 05.036Z"}",
      "method"=>"GET"
    }
  ]
  */
  /* eslint-enable */


  // --- the test record object for parsing
  const record = {
    'agent': {
      'type': 'fluentbit',
    },
    'azp': 'chefs-frontend',
    'contentLength': '323',
    'event': {
      'type': 'http',
      'module': 'node.css',
      'kind': 'event',
      'dataset': 'node.css.http',
    },
    'hostname': 'chefs-app-123',
    'httpVersion': '1.1',
    'ip': '::ffff:10.97.8.1',
    'level': 'http',
    'log': {
      'responseTime': 624,
      'azp': 'chefs-frontend',
      'contentLength': '323',
      'httpVersion': '1.1',
      'ip': ': :ffff: 10.97.8.1',
      'method': 'GET',
      'path': '/api/v1/forms',
      'statusCode': 200,
      'userAgent': 'Mozilla/5.0',
      'level': 'http',
      'message': 'GET /api/v1/forms 200 624ms', 'timestamp': '2021-07-14T23: 17: 05.036Z',
    },
    'logFilePath': '/var/log/app.log',
    'logFileOffset': 759,
    'message': 'GET /api/v1/forms 200 624ms',
    'method': 'GET',
    'namespace': '123-dev',
    'path': '/api/v1/forms',
    'product': 'chefs',
    'responseTime': 624,
    'statusCode': 200,
    'timestamp': '2021-07-14T23:17:05.036Z',
    'userAgent': 'Mozilla/5.0',
  };

  // expect the parser's match() method to return true and run for this test record
  expect(filter.matches(record)).toEqual(true);

  // --- run the parser:
  filter.apply(record);

  // --- test the result:
  // contentLength
  expect(record).not.toHaveProperty('contentLength');
  expect(record).toHaveProperty('http.response.body.bytes', '323');
  // hostname
  expect(record).not.toHaveProperty('hostname');
  expect(record).toHaveProperty('host.hostname', 'chefs-app-123');
  // httpVersion
  expect(record).not.toHaveProperty('httpVersion');
  expect(record).toHaveProperty('http.version', '1.1');
  // ip
  expect(record).not.toHaveProperty('ip');
  expect(record).toHaveProperty('source.address', '10.97.8.1');
  // log
  expect(record).toHaveProperty('event.original.responseTime', 624);
  // logFileOffset
  expect(record).not.toHaveProperty('logFileOffset');
  expect(record).toHaveProperty('log.file.offset', 759);
  // logFilePath
  expect(record).not.toHaveProperty('logFilePath');
  expect(record).toHaveProperty('log.file.path', '/var/log/app.log');
  // method
  expect(record).not.toHaveProperty('method');
  expect(record).toHaveProperty('http.request.method', 'GET');
  // namespace
  expect(record).not.toHaveProperty('namespace');
  expect(record).toHaveProperty('host.namespace', '123-dev');
  // path
  expect(record).not.toHaveProperty('path');
  expect(record).toHaveProperty('url.original', '/api/v1/forms');
  // responseTime
  expect(record).not.toHaveProperty('responseTime');
  expect(record).toHaveProperty('http.response.time', 624);
  // statusCode
  expect(record).not.toHaveProperty('statusCode');
  expect(record).toHaveProperty('http.response.status_code', 200);
  // timestamp
  expect(record).not.toHaveProperty('timestamp');
  expect(record).toHaveProperty('@timestamp', '2021-07-14T23:17:05.036Z');
  // userAgent
  expect(record).not.toHaveProperty('userAgent');
  expect(record).toHaveProperty('user_agent.original', 'Mozilla/5.0');

  // fields already in ECS format:
  // message

  // extra fields added on top of ECS:
  // azp, level, product

  // use snapshot
  expect(record).toMatchSnapshot('2f015294-45f8-4235-949b-667236bea641');
});
