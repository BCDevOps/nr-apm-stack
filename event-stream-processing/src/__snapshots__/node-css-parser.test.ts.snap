// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic: 2f015294-45f8-4235-949b-667236bea641 1`] = `
Object {
  "@timestamp": "2021-07-14T23:17:05.036Z",
  "agent": Object {
    "type": "fluentbit",
  },
  "azp": "chefs-frontend",
  "event": Object {
    "dataset": "node.css.http",
    "kind": "event",
    "module": "node.css",
    "original": Object {
      "azp": "chefs-frontend",
      "contentLength": "323",
      "httpVersion": "1.1",
      "ip": ": :ffff: 10.97.8.1",
      "level": "http",
      "message": "GET /api/v1/forms 200 624ms",
      "method": "GET",
      "path": "/api/v1/forms",
      "responseTime": 624,
      "statusCode": 200,
      "timestamp": "2021-07-14T23: 17: 05.036Z",
      "userAgent": "Mozilla/5.0",
    },
    "type": "http",
  },
  "host": Object {
    "hostname": "chefs-app-123",
    "namespace": "123-dev",
  },
  "http": Object {
    "request": Object {
      "method": "GET",
    },
    "response": Object {
      "body": Object {
        "bytes": "323",
      },
      "status_code": 200,
      "time": 624,
    },
    "version": "1.1",
  },
  "level": "http",
  "log": Object {
    "file": Object {
      "offset": 759,
      "path": "/var/log/app.log",
    },
  },
  "message": "GET /api/v1/forms 200 624ms",
  "product": "chefs",
  "source": Object {
    "address": "10.97.8.1",
  },
  "url": Object {
    "original": "/api/v1/forms",
  },
  "user_agent": Object {
    "original": "Mozilla/5.0",
  },
}
`;
